### ISSUE TITLE
Corrupt KTX export for decompressed HDR data

### ISSUE THREAD

**Issue by rovedit** [ARM-software/astc-encoder#352]
**Problem:**
When I compress a 3D Texture in RGBA 32bit Float format into ASTC and afterwards decompress it, the header is corrupt thus the file cannot be read anymore.

**Process:**
In detail I proceded as follows:

1. Compress a 3D texture in RGBA Float (32bit) format with the astcencoder like this:
`astcenc-avx2.exe -cH ctbl3d_rgba.ktx ctbl3d_astc.ktx 4x4 -medium`. 
2. Decompress it again like this:
`astcenc-avx2.exe -dH ctbl3d_astc.ktx ctbl3d_decoded.ktx`
3. Try to read `ctbl3d_decoded.ktx` using [Khronos' offical KTX library](https://github.com/KhronosGroup/KTX-Software)

The output of [Khronos' ktxinfo](https://github.khronos.org/KTX-Software/ktxtools/ktxinfo.html) tool for `ctbl3d_decoded.ktx` looks like this:
<details>

```
Header

identifier: «KTX 11»\r\n\x1A\n
endianness: 0x4030201
glType: 0x140b
glTypeSize: 2
glFormat: 0x1908
glInternalformat: 0x1908
glBaseInternalformat: 0x1908
pixelWidth: 130
pixelHeight: 384
pixelDepth: 384
numberOfArrayElements: 0
numberOfFaces: 1
numberOfMipLevels: 1
bytesOfKeyValueData: 0
The KTX 1 file pHeader is invalid:
  it has invalid data such as bad glTypSize, improper dimensions,
improper number of faces or too many levels.
```

</details>

**Possible reason:**
The ASTC encoder uses a forbidden `glFormat` and `glInternalformat` combination in the header when saving decompressed HDR data in KTX.
The problematic code in the Khronos library is:
```c++
if (pHeader->glFormat == pHeader->glInternalformat) {
        // glInternalFormat is either unsized (which is no longer and should
        // never have been supported by libktx) or glFormat is sized.
        return KTX_FILE_DATA_ERROR;
    }
```

**Possible solution:**
On inspection, `glFormat` and `glInternalformat` both are valued `GL_RGBA` (decimal 6408), though `glInternalformat` should be sized. Further inspecting, the astcencoder seems to save the decompressed data as 16bit HalfFloat, thus `GL_RGBA16F` (decimal 34842) seems fitting for the sized format.
Even better would be, if the decompression output can be tuned to be also 32bit Float (`GL_RGBA32F`) as this is natively readable in other applications.

**Comment by solidpixel** [comment]
Is it possible to share the reproducing texture for this one? I'm not sure I have a 32-bit per component 3D texture to hand to test.

**Comment by rovedit** [comment]
Thanks for your fast answer. The 3D texture is originally a vector field, openly available at [ETH Zürich](https://cgl.ethz.ch/research/visualization/data.php) (Cloud-topped Boundary Layer).

Here is a smaller subset of the file (I reduced depth to 16) in KTX format. I also just reproduced the error for the test file.
[3D_Texture.zip](https://github.com/ARM-software/astc-encoder/files/9261395/3D_Texture.zip)

**Comment by rovedit** [comment]
I also edited the code of the astc-encoder a bit. I think I could eliminate the problem editing `Source\astcenccli_image_load_store.cpp`:

<details>

```c++
static bool store_ktx_uncompressed_image(
	const astcenc_image* img,
	const char* filename,
	int y_flip
) {
	unsigned int dim_x = img->dim_x;
	unsigned int dim_y = img->dim_y;
	unsigned int dim_z = img->dim_z;

	int bitness = img->data_type == ASTCENC_TYPE_U8 ? 8 : 16;
	int image_components = determine_image_components(img);

	ktx_header hdr;

	static const int gl_format_of_components[4] {
		GL_LUMINANCE, GL_LUMINANCE_ALPHA, GL_RGB, GL_RGBA
	};

	// Fix
	static const int gl_internalformat_of_components[16]{
		GL_R8, GL_RG8, GL_RGB8, GL_RGBA8,
		GL_R16F, GL_RG16F, GL_RGB16F, GL_RGBA16F,
		GL_R32F, GL_RG32F, GL_RGB32F, GL_RGBA32F
	};
	// Fix (only for Half Float, not for Float32)
	int bitness_level = (bitness == 8) ? 0 : 4;

	memcpy(hdr.magic, ktx_magic, 12);
	hdr.endianness = 0x04030201;
	hdr.gl_type = (bitness == 16) ? GL_HALF_FLOAT : GL_UNSIGNED_BYTE;
	hdr.gl_type_size = bitness / 8;
	hdr.gl_format = gl_format_of_components[image_components - 1];
	// Fix
	hdr.gl_internal_format = gl_internalformat_of_components[bitness_level + (image_components - 1)];
	// Old
	//hdr.gl_internal_format = gl_format_of_components[image_components - 1];
	hdr.gl_base_internal_format = gl_format_of_components[image_components - 1];
```

</details>

... and `Source\astcenccli_image_load_store.cpp`:
<details>

```c++
// Adding the needed enums
#define GL_R8																			  0x8229
#define GL_R16F																			0x822D
#define GL_R32F																			0x822E
#define GL_RG8																		  0x822B
#define GL_RG16F																		0x822F
#define GL_RG32F																		0x8230
#define GL_RGB8																		  0x8051
#define GL_RGB16F																		0x881B
#define GL_RGB32F																		0x8815
#define GL_RGBA8																		0x8058
#define GL_RGBA16F																	0x881A
#define GL_RGBA32F																  0x8814
#define GL_LUMINANCE8                               0x8040
#define GL_LUMINANCE16                              0x8042
#define GL_LUMINANCE32F_ARB                         0x8818
#define GL_LUMINANCE8_ALPHA8	  									  0x8045
#define GL_LUMINANCE8_ALPHA16 										  0x8048
#define GL_LUMINANCE_ALPHA32F_ARB 								  0x8819
```

</details>

**Comment by solidpixel** [comment]
It looks like this can be reproduced with any output uncompressed KTX texture - 2D LDR fails in the same way. 

```
./astcenc-avx2 -cl in.png out.astc 6x6 -medium
./astcenc-avx2 -dl out.astc out.ktx
./ktxinfo out.ktx

The KTX 1 file pHeader is invalid:
  it has invalid data such as bad glTypSize, improper dimensions,
improper number of faces or too many levels.
```

**Comment by solidpixel** [comment]
I've split off the request for 32-bit output into a separate issue, so this one can focus on the bug fix for the existing code.

**Comment by solidpixel** [comment]
Does this one work for you: #354 (tested here on 2D images that I have a viewer for and it seems to solve the problems, and ktxinfo is happy)

**Comment by rovedit** [comment]
Sorry for the late answer, I got ill and was confined to bed not able to test anything.

I did test your changes and they do seem to do the trick. Thanks for the really fast fix! But I'm still rendering only a black picture when I load the exported data. My local testing suggests it has to do with the input data type:

When I compress 32bit floats, decompress it and finally load it into my application it renders a black picture.
When I do the same process with 16bit float data (using an [IEEE fp16 library](https://github.com/Maratyszcza/FP16) for conversion), the content is rendered correctly.

I checked R/RG/RGB/RGBA test data which should only render a green image (or red for R) and all exhibit the above described phenomenon. Here is the test data I used:
[fp16.zip](https://github.com/ARM-software/astc-encoder/files/9281624/fp16.zip)
[fp32.zip](https://github.com/ARM-software/astc-encoder/files/9281625/fp32.zip)
